### PR TITLE
[c++] fix Booster.refit() segfault with linear_tree on single-class data

### LINKED ISSUE #6792: [python-package] `refit()` segfaults when label only contains 1 class
## Description

Using the Python package, when fitting a binary classification model with only categorical features and using linear trees, `refit()` segfaults if `label` has only 1 class.

## Reproducible example

```python
import lightgbm as lgb
import numpy as np

X = np.random.randint(low=0, high=10, size=(1_000, 5))
y = np.ones(shape=(X.shape[0],))

dtrain = lgb.Dataset(
  data=X,
  label=y,
  categorical_feature=list(range(X.shape[1]))
)

bst = lgb.train(
  train_set=dtrain,
  params={
    "objective": "binary",
    "num_leaves": 7,
    "linear_tree": True
  },
  num_boost_round=2
).refit(X, label=y, categorical_feature=list(range(X.shape[1])))
```

Logs from `refit()`:

```text
[LightGBM] [Warning] Contains only one class
[LightGBM] [Info] Number of positive: 1000, number of negative: 0
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000032 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 55
[LightGBM] [Info] Number of data points in the train set: 1000, number of used features: 5
Segmentation fault: 11
```

Training the same model but without `linear_tree=True`, `refit()` does not segfault or otherwise raise an exception... it just trains a stub model, as expected.

```python
bst = lgb.train(
  train_set=dtrain,
  params={
    "objective": "binary",
    "num_leaves": 7,
  },
  num_boost_round=2
)
bst.refit(X, label=y, categorical_feature=list(range(X.shape[1])))
```

Logs from `refit()`:

```text
[LightGBM] [Warning] Contains only one class
[LightGBM] [Info] Number of positive: 1000, number of negative: 0
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000035 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 55
[LightGBM] [Info] Number of data points in the train set: 1000, number of used features: 5
<lightgbm.basic.Booster object at 0x152a19690>
```

Tree content:

```python
bst.dump_model()["tree_info"]
```

```json
[{"tree_index": 0, "num_leaves": 1, "num_cat": 0, "shrinkage": 1, "tree_structure": {"leaf_value": 34.53957599234088, "leaf_count": 1000}}]
```

## Environment info

LightGBM version or commit hash: https://github.com/microsoft/LightGBM/commit/226e7f7de8358161072524f4f91ed8366a1dead6

Command(s) you used to install LightGBM

```shell
cmake -B build -S . -DUSE_OPENMP=OFF
cmake --build build --target _lightgbm -j4
sh build-python.sh install --precompile
```

## Additional Comments

Fitting a model with a single class is maybe not that useful of a case for LightGBM, but still... LightGBM should **never** do something that results in process-crashing segmentation fault.

And these sorts of situations can occur unintentionally, for example in AutoML systems where random subsets of training data are selected.

### PR DESCRIPTION
**Pull request by anishkb-dev**
## Summary

Fixes #6792 — `Booster.refit()` segfaults when the training label contains only one class and the model was trained with `linear_tree=True`.

## Root cause

When the label has a single class, the objective decides that class doesn't need boosting and LightGBM emits a plain constant tree, built through the `Tree(max_leaves, false, /*is_linear=*/false)` placeholder. Such a tree never has its per-leaf linear-model arrays (`leaf_features_`, `leaf_coeff_`, `leaf_const_`) allocated.

`LinearTreeLearner` treated those trees as linear regardless, giving two out-of-bounds reads with the same cause:

- `CalculateLinear()` unconditionally calls `SetIsLinear(true)`, then on the `is_refit` path reads `tree->LeafFeatures(i)` — indexing past the end of a zero-length vector.
- `AddPredictionToScore()` reads `LeafConst`/`LeafCoeffs` for the same trees.

Either one is enough to segfault the process during `refit()`.

## Fix

Check whether the tree is actually linear and fall back to the plain non-linear handling when it isn't, so a constant tree stays constant.

Keeping these trees non-linear also preserves the `(num_leaves <= 1 && !is_linear)` early return in `Tree`'s model-string constructor. Flagging them linear instead makes the deserializer parse the legitimately-empty `leaf_weight` field and abort with an uncaught `std::runtime_error`.

## Relation to #7378

#7378 targets the same issue by marking these trees linear and allocating the missing arrays. That path runs into the model-string problem described above: the resulting model aborts on deserialization rather than segfaulting, which trades one crash for another that is harder to attribute.

This PR takes the opposite direction — leave the constant tree non-linear — which keeps `model_to_string()` round-tripping. Happy to defer to #7378 or combine the two if maintainers prefer that shape; flagging the interaction so it isn't discovered after a merge.

## Testing

Added a regression test in `tests/python_package_test/test_engine.py` that refits a `linear_tree` model on single-class data and round-trips the result through `model_to_string()`, covering both the segfault and the deserialization abort.

Full Python suite on this branch: **1355 passed, 8 skipped, 12 xfailed, 4 xpassed**, matching the clean-tree baseline.

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thank you for your interest in LightGBM, but no. We already have 2 other PRs trying to address this (#7233 and #7378).

Don't blast AI-generated PRs at projects you care about, please. Consider reading this: https://blog.probabl.ai/maintaining-open-source-age-of-gen-ai

You are blocked from LightGBM for 7 days. After that, you're welcome to come back and make other contributions but please:

* write the PR description, title, and review comments yourself
* do not open PRs for issues that already have in-progress PRs
* be sure you understand the content of your PR and have reviewed it yourself before opening.
